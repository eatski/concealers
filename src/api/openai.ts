import { OpenAI, APIError } from 'openai'

import type { Character } from '../shared'

export interface CharacterResponse {
  nextStatement: string
  urgency: number
}

function generatePromptForCharacter(currentCharacter: Character, allCharacters: Character[]) {
  const otherCharacters = allCharacters.filter(char => char !== currentCharacter)
  
  return `
あなたは「${currentCharacter.name}」というキャラクターです。

あなたの情報:
名前: ${currentCharacter.name}
説明: ${currentCharacter.description}
隠された情報: ${currentCharacter.hiddenPrompt}

他のキャラクター:
${otherCharacters.map(char => `
名前: ${char.name}
説明: ${char.description}
`).join('\n')}

この状況で、あなたが次に発言したいことを決定してください。
`.trim()
}

export async function sendRequest(apiKey: string, characters: Character[]): Promise<CharacterResponse[]> {
  if (!apiKey) return []
  
  const openai = new OpenAI({
    apiKey: apiKey,
    dangerouslyAllowBrowser: true
  })

  try {
    const responses = await Promise.all(
      characters.map(async (character) => {
        const prompt = generatePromptForCharacter(character, characters)
        
        const completion = await openai.chat.completions.create({
          model: 'gpt-4',
          messages: [
            { 
              role: 'system', 
              content: 'キャラクターの次の発言を決定し、関数を呼び出して回答してください。' 
            },
            { 
              role: 'user', 
              content: prompt 
            }
          ],
          functions: [
            {
              name: 'decideNextStatement',
              description: 'キャラクターの次の発言を決定する',
              parameters: {
                type: 'object',
                properties: {
                  nextStatement: {
                    type: 'string',
                    description: '次に発言しようとしていることを100文字程度で説明'
                  },
                  urgency: {
                    type: 'integer',
                    description: '発言への意欲（1: 消極的, 5: 積極的）',
                    minimum: 1,
                    maximum: 5
                  }
                },
                required: ['nextStatement', 'urgency']
              }
            }
          ],
          function_call: { name: 'decideNextStatement' }
        })

        const functionCall = completion.choices[0].message.function_call
        if (!functionCall || !functionCall.arguments) {
          throw new Error('APIからの応答が不正です')
        }

        try {
          const response = JSON.parse(functionCall.arguments) as CharacterResponse
          return response
        } catch (e) {
          console.error('JSON解析エラー:', e)
          throw new Error('APIからの応答の解析に失敗しました')
        }
      })
    )

    return responses
  } catch (error) {
    console.error('Error:', error)
    if (error instanceof APIError) {
      if (error.status === 401) {
        throw new Error('APIキーが無効です')
      } else if (error.status === 429) {
        throw new Error('リクエスト制限に達しました')
      }
    }
    throw error
  }
}