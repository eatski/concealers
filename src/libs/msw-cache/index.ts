import { unlink } from "node:fs/promises";
import {
  http,
  type DefaultBodyType,
  HttpResponse,
  type StrictRequest,
  bypass,
} from "msw";
import { setupServer } from "msw/node";
import { FileSystemCache } from "../file-system-cache";

interface CacheData {
  request: Record<string, unknown>;
  response: unknown;
}

class ResponseCache {
  private readonly usedCachePath = new Set<string>();
  constructor(private readonly cache: FileSystemCache) {}

  private static async configToRequestRecord(
    config: StrictRequest<DefaultBodyType>,
  ): Promise<Record<string, unknown>> {
    return {
      url: config.url,
      body: await config.text(),
      method: config.method,
    };
  }

  public async get(
    configForKey: StrictRequest<DefaultBodyType>,
  ): Promise<unknown> {
    const record = await ResponseCache.configToRequestRecord(configForKey);
    const key = JSON.stringify(record);
    this.usedCachePath.add(this.cache.getPath(key));
    const cached = await this.cache.get<CacheData>(key);
    return cached?.response;
  }

  public async set(
    configForKey: StrictRequest<DefaultBodyType>,
    value: unknown,
  ) {
    const record = await ResponseCache.configToRequestRecord(configForKey);
    const key = JSON.stringify(record);
    this.usedCachePath.add(this.cache.getPath(key));
    await this.cache.set(key, {
      request: record,
      response: value,
    });
  }

  public async clearUnusedCache() {
    const caches = await this.cache.load();
    const pathes = caches.files
      .map((file) => file.path)
      .filter((path) => !this.usedCachePath.has(path));
    await Promise.all(pathes.map(unlink));
  }
}

export const initMswCacheServer = (basePath: string) => {
  const cache = new ResponseCache(
    new FileSystemCache({
      basePath: basePath,
    }),
  );
  const server = setupServer(
    http.all("*", async ({request}) => {
      const cached = await cache.get(request.clone());
      if (typeof cached === "object") {
        return HttpResponse.json(cached);
      }
      console.log("Cache not hit. ", request.url);
    	const response = await fetch(bypass(request));
      if (response.status === 200) {
        const text = await response.clone().json();
        await cache.set(request, text);
      }
      return HttpResponse.json(await response.json());
    }),
  );
  return {
    listen: () => server.listen(),
    close: () => {
      server.close();
      return cache.clearUnusedCache();
    },
  };
};
