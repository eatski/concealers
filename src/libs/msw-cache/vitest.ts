import path from "node:path";
import filenamify from "filenamify";
import { afterAll, beforeAll } from "vitest";
import { initMswCacheServer } from ".";

export const applyTestHooks = () => {
	let server: ReturnType<typeof initMswCacheServer> | null = null;
	beforeAll(({ name, file }) => {
		if (!file?.name) {
			throw new Error("suite.file.name is not defined");
		}
		const parserdPath = path.parse(file.name);
		const filename = filenamify(name, { replacement: "_" });
		server = initMswCacheServer(
			path.resolve(
				process.cwd(),
				parserdPath.dir,
				"__data__",
				parserdPath.name + parserdPath.ext,
				filename,
			),
		);
		server.listen();
	});
	afterAll(async ({ tasks }) => {
		if (tasks.every((t) => t.result?.state === "pass")) {
			await server?.close();
		}
	});
};
