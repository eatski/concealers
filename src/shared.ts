export interface Character {
  name: string
  description: string
  hiddenPrompt: string
}

export interface GameSettings {
  commonPrompt: string
  characters: Character[]
}

export interface MemoryItem {
  recognizedInfo: string
  thought: string
  tags: string[]
}

export interface CharacterMemories {
  characterName: string
  memories: MemoryItem[]
  urgency: 1 | 2 | 3
}

export interface CharacterSpeech {
  characterName: string
  speech: string
}

export interface RoutineResult {
  characterMemories: CharacterMemories[]
  speech: CharacterSpeech | null
}