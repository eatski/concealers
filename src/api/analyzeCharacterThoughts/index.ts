import { OpenAI } from 'openai'
import { z } from 'zod'
import type { Character, CharacterMemories, MemoryItem, RoutineResult } from '../../shared'
import { buildPrompt } from '../../libs/prompt-builder'
import { makeOpenAIRequest } from '../../libs/openai-request'

export interface CreateCharacterThoughtsArgs {
  openai: OpenAI
  commonPrompt: string
  currentCharacter: Character
  allCharacters: Character[]
  history: RoutineResult[]
  relevantMemories: MemoryItem[]
}

interface CreateCharacterAnalysisPromptArgs {
  currentCharacter: Character
  allCharacters: Character[]
  commonPrompt: string
  history: RoutineResult[]
  relevantMemories: MemoryItem[]
}

// 内部的に使用する型定義
export interface CharacterMemoriesWithUrgency extends CharacterMemories {
  urgency: 1 | 2 | 3
}

// zodスキーマ
const memoryItemResponseSchema = z.object({
  recognizedInfo: z.string().describe('現在の状況で認識した情報を100文字程度で説明'),
  thought: z.string().describe('その状況に対して考えたことを100文字程度で説明'),
  tags: z.array(z.string()).describe('この記憶に関連するキーワードやキャラクター名の配列(1~2個)')
})

const characterThoughtsResponseSchema = z.object({
  memories: z.array(memoryItemResponseSchema).describe('認識した情報とそれに対する考えの配列'),
  urgency: z.union([z.literal(1), z.literal(2), z.literal(3)])
    .describe('全体的な発言への意欲（1: 特に話すことがない・誰かが話すのを聞きたい・直近発言したばかり, 2: 強い意欲があるわけではないが話すことがある, 3: 積極的に話したい・話す責務がある）')
})

function createCharacterAnalysisPrompt({
  currentCharacter,
  allCharacters,
  commonPrompt,
  history,
  relevantMemories
}: CreateCharacterAnalysisPromptArgs) {
  const otherCharacters = allCharacters.filter(char => char !== currentCharacter)
  
  const sections = [
    {
      name: '共通の情報',
      content: commonPrompt
    },
    {
      name: 'あなたの情報',
      content: `名前: ${currentCharacter.name}
説明: ${currentCharacter.description}
隠している情報: ${currentCharacter.hiddenPrompt}`
    },
    {
      name: '他のキャラクター',
      content: otherCharacters.map(char =>
        `名前: ${char.name}
説明: ${char.description}`
      ).join('\n')
    }
  ]

  if (relevantMemories.length > 0) {
    sections.push({
      name: '関連する過去の記憶',
      content: relevantMemories.map(memory =>
        `認識した情報: ${memory.recognizedInfo}
考え: ${memory.thought}`
      ).join('\n')
    })
  }

  if (history.length > 0) {
    sections.push({
      name: '直近の会話',
      content: history.slice(-3).map((routine) =>
        `${routine.speech
          ? `${routine.speech.characterName === currentCharacter.name ? "あなた" : routine.speech.characterName}の発言: ${routine.speech.speech}`
          : '発言なし'}`
      ).join('\n')
    })
  }

  return buildPrompt(sections)
}

export async function createCharacterThoughts({
  openai,
  commonPrompt,
  currentCharacter,
  allCharacters,
  history,
  relevantMemories
}: CreateCharacterThoughtsArgs): Promise<CharacterMemoriesWithUrgency> {
  const prompt = createCharacterAnalysisPrompt({
    currentCharacter,
    allCharacters,
    commonPrompt,
    history,
    relevantMemories
  })

  const response = await makeOpenAIRequest({
    openai,
    schema: characterThoughtsResponseSchema,
    systemPrompt: '現在の状況から複数の観点で状況を認識し、それぞれに対する考えを回答してください。各認識に関連するキーワードやキャラクター名をタグとして付与してください。また、全体的な発言意欲も回答してください。過去あなたの発言が多い場合は発言意欲は低めにしてください。',
    userPrompt: prompt,
    functionName: 'analyzeThoughtsAndUrgency',
    functionDescription: 'キャラクターの状況認識と考え、全体的な発言意欲'
  })

  return {
    characterName: currentCharacter.name,
    memories: response.memories,
    urgency: response.urgency
  }
}