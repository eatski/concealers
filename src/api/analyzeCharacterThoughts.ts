import { OpenAI, APIError } from 'openai'

import type { Character } from '../shared'

export interface CharacterThought {
  thought: string
  urgency: 1 | 2 | 3
}

import { type RoutineResult } from '../shared'

function createCharacterAnalysisPrompt(
  currentCharacter: Character,
  allCharacters: Character[],
  commonPrompt: string,
  history: RoutineResult[]
) {
  const otherCharacters = allCharacters.filter(char => char !== currentCharacter)
  
  return `
<共通の情報>
${commonPrompt}
</共通の情報>
<あなたの情報>
名前: ${currentCharacter.name}
説明: ${currentCharacter.description}
隠している情報: ${currentCharacter.hiddenPrompt}
</あなたの情報>
<他のキャラクター>
${otherCharacters.map(char => `
名前: ${char.name}
説明: ${char.description}
`).join('\n')}
${history.length > 0
  ? `
</他のキャラクター>
<これまでの会話>
${history.map((routine) => `
${routine.thoughts
.filter(thought => thought.characterName === currentCharacter.name)
.map(thought => `あなたの考え: ${thought.thought}`).join('\n')}
${routine.speech
? `${routine.speech.characterName === currentCharacter.name ? "あなた" : routine.speech.characterName}の発言: ${routine.speech.speech}`
: '発言なし'}`
).join('\n')}
`
  : ''}
</これまでの会話>
`.trim()
}

export async function createCharacterThoughts(
  openai: OpenAI,
  commonPrompt: string,
  characters: Character[],
  history: RoutineResult[]
): Promise<CharacterThought[]> {
  if (characters.length === 0) return []

  try {
    const responses = await Promise.all(
      characters.map(async (character) => {
        const prompt = createCharacterAnalysisPrompt(character, characters, commonPrompt, history)
        
        const completion = await openai.chat.completions.create({
          model: "gpt-4o",
          messages: [
            { 
              role: 'system',
              content: 'あなたの現在の心情と発言意欲を回答してください。過去あなたの発言が多い場合は発言意欲は低めにしてください。'
            },
            { 
              role: 'user', 
              content: prompt 
            }
          ],
          functions: [
            {
              name: 'analyzeThoughtAndUrgency',
              description: 'キャラクターの考えと発言意欲',
              parameters: {
                type: 'object',
                properties: {
                  thought: {
                    type: 'string',
                    description: '現在考えていることを100文字程度で説明'
                  },
                  urgency: {
                    type: 'integer',
                    description: '発言への意欲（1: 特に話すことがない・誰かが話すのを聞きたい・直近発言したばかり, 2: 強い意欲があるわけではないが話すことがある, 3: 積極的に話したい・話す責務がある）',
                    minimum: 1,
                    maximum: 3
                  }
                },
                required: ['thought', 'urgency']
              }
            }
          ],
          function_call: { name: 'analyzeThoughtAndUrgency' }
        })

        const functionCall = completion.choices[0].message.function_call
        if (!functionCall || !functionCall.arguments) {
          throw new Error('APIからの応答が不正です')
        }

        try {
          const response = JSON.parse(functionCall.arguments) as CharacterThought
          return response
        } catch (e) {
          console.error('JSON解析エラー:', e)
          throw new Error('APIからの応答の解析に失敗しました')
        }
      })
    )

    return responses
  } catch (error) {
    console.error('Error:', error)
    if (error instanceof APIError) {
      if (error.status === 401) {
        throw new Error('APIキーが無効です')
      } else if (error.status === 429) {
        throw new Error('リクエスト制限に達しました')
      }
    }
    throw error
  }
}