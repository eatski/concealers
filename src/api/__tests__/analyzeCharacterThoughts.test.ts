import { describe, it, expect } from 'vitest'
import { OpenAI } from 'openai'
import { createCharacterThoughts } from '../analyzeCharacterThoughts'
import { applyTestHooks } from '../../libs/msw-cache/vitest'
import { type Character, type RoutineResult } from '../../shared'

const commonPrompt = `
## 概要
あなたたちは探偵事務所のメンバーです。
最近、街で連続強盗事件が発生しており、その捜査を依頼されました。
事件の真相を探るため、それぞれが持つ情報を共有し合う必要があります。

## 基本ルール
- 各自が持つ情報は慎重に扱う必要があります
- 相手の反応を見ながら、適切なタイミングで情報を開示してください
- 他のメンバーの発言から、真相のヒントを見つけ出してください
`.trim()

const characters: Character[] = [
  {
    name: '佐藤刑事',
    description: 'ベテラン刑事。冷静な判断力と鋭い洞察力を持つ。',
    hiddenPrompt: '実は犯人の手口から、これが内部犯行である可能性が高いと気づいている。'
  },
  {
    name: '山田探偵',
    description: '新進気鋭の私立探偵。独自の情報網を持つ。',
    hiddenPrompt: '犯行現場付近で不審な警備員の姿を目撃している。'
  }
]

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
})

describe('createCharacterThoughts', () => {
  applyTestHooks()

  it('履歴がない状態でのキャラクターの思考を生成できる', async () => {
    const currentCharacter = characters[0]
    const history: RoutineResult[] = []

    const result = await createCharacterThoughts({
      openai,
      commonPrompt,
      currentCharacter,
      allCharacters: characters,
      history,
      relevantMemories: []
    })

    expect(result).toMatchSnapshot()
  })

  it('履歴が3件ある状態でのキャラクターの思考を生成できる', async () => {
    const currentCharacter = characters[0]
    const history: RoutineResult[] = [
      {
        characterMemories: [{
          characterName: characters[1].name,
          memories: []
        }],
        speech: {
          characterName: characters[1].name,
          speech: '現場周辺の警備体制に不審な点がありました。'
        }
      },
      {
        characterMemories: [{
          characterName: characters[1].name,
          memories: []
        }],
        speech: {
          characterName: characters[1].name,
          speech: '特に夜間の警備に空白の時間帯があったようです。'
        }
      },
      {
        characterMemories: [{
          characterName: characters[1].name,
          memories: []
        }],
        speech: {
          characterName: characters[1].name,
          speech: 'その時間帯に不審な人物の出入りがあったという目撃情報もあります。'
        }
      }
    ]

    const result = await createCharacterThoughts({
      openai,
      commonPrompt,
      currentCharacter,
      allCharacters: characters,
      history,
      relevantMemories: []
    })

    expect(result).toMatchSnapshot()
  })
})