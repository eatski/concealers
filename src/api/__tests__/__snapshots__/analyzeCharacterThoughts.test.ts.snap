// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createCharacterThoughts > 履歴が3件ある状態でのキャラクターの思考を生成できる 1`] = `
{
  "characterName": "佐藤刑事",
  "memories": [
    {
      "recognizedInfo": "山田探偵が現場周辺の警備体制に不審な点を指摘しています。特に夜間の警備に空白時間があったとのこと。",
      "tags": [
        "警備体制",
        "内部犯行",
      ],
      "thought": "もし内部犯行なら、警備に隙が生じていたのは手引きによるものか？",
    },
    {
      "recognizedInfo": "夜間の空白時間帯に不審な人物の出入りが目撃されています。",
      "tags": [
        "アリバイ確認",
        "目撃情報",
      ],
      "thought": "内部犯行が疑われる状況なので、関係者のアリバイを確認する必要がある。",
    },
  ],
  "urgency": 1,
}
`;

exports[`createCharacterThoughts > 履歴がない状態でのキャラクターの思考を生成できる 1`] = `
{
  "characterName": "佐藤刑事",
  "memories": [
    {
      "recognizedInfo": "連続強盗事件が発生しており、探偵事務所のメンバーで捜査することになった。",
      "tags": [
        "連続強盗事件",
        "捜査チーム",
      ],
      "thought": "事件解決に向けてチームで情報を共有し、連携を深めることが重要。",
    },
    {
      "recognizedInfo": "犯人の手口から内部犯行の可能性が高いと考えている。",
      "tags": [
        "犯人手口",
        "内部犯行",
      ],
      "thought": "この情報をどう活用するか、開示のタイミングを慎重に見極める必要がある。",
    },
  ],
  "urgency": 2,
}
`;
